proposals/devcontainer-templates.md

```diff
@@ -0,0 +1,144 @@
+# Dev Container Templates reference
+
+Development container "Templates" are self-contained, shareable units of development container configuration. The name comes from the idea that adding a Template to your project helps it get up and running with a containerized environment. The Templates describe the appropriate container image, [Features](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md), runtime arguments for starting the container, and VS Code extensions that should be installed. Each provides a container configuration file ([`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson)) and other needed files that you can drop into any existing folder as a starting point for containerizing your project.
+
+Template metadata is captured by a `devcontainer-template.json` file in the root folder of the Template.
+
+## Folder structure 
+
+A Template is a self contained entity in a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.
+
+```
++-- template
+|    +-- devcontainer-template.json
+|    +-- .devcontainer
+|        +-- devcontainer.json 
+|        +-- Dockerfile (optional)
+|        +-- docker-compose.yml (optional)
+|        +-- (other files)
+|    +-- (other files)
+```
+
+## devcontainer-template.json properties
+
+The `devcontainer-template.json` file defines information about the Template to be used by any [supporting tools](#template-supporting-tools-and-services).
+
+The properties of the file are as follows:
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| `id` | string | ID of the Template. The `id` should be unique in the context of the repository/published package where the Template exists and must match the name of the directory where the `devcontainer-template.json` resides. |
+| `version` | string | The semantic version of the Template. |
+| `name` | string | Name of the Template. |
+| `description` | string | Description of the Template. |
+| `documentationURL` | string | Url that points to the documentation of the Template. |
+| `licenseURL` | string | Url that points to the license of the Template. |
+| `type` | string | Type of the dev container (image, dockerfile, dockerCompose) created by the Template. |
```

Comment:
How is this used?

Reply:
Currently, the `definitions` have this `type` property (`singleContainer` and `dockerCompose`) which is used by the UI to determine how to start or stop the container and configure few other things. Hence, added it here. 

Also, as per today's discussion, this will also be useful for listing Templates and people can sort with the `type`

Reply:
Perhaps the tooling could add some of these attributes.  I imagine from some `devcontainer template package` command, it would not be too difficult to programatically add several of these attributes (and ensures consistency/doesn't require action from the author).

Reply:
Just wondering what if the tooling mistags some times ?

Example - lets say a Template contains all of three files (.devcontainer.json, dockerfile and dockercompose), the Template author writes code in such a way that the dev container solely depends on an image and has only added boiler plate code for `dockerfile` and `dockerCompose` or samples of how to configure using them. If the tooling simply determines the value based on the file presence of some other attributes, I wonder it might mis configure sometimes 🤔

Reply:
Makes sense. Removed `type`

Reply:
Sorry, I wasn't clear: I suggest to keep `type` and we can add tooling support later (whenever we get to it). :)